File: web/src/crypto/account.ts
// Account keys, password-derived keys and the encrypted key bundle,
// PROTOCOL.md §3.
import { concat, equal, randomBytes, utf8Encode } from './bytes';
import { aeadDecrypt, aeadEncrypt, argon2idHash, ed25519Public, type KdfParams, NONCE_SIZE, sign, TAG_SIZE, x25519Public } from './primitives';

export interface AccountKeys {
  signSeed: Uint8Array;
  signPub: Uint8Array;
  encPriv: Uint8Array;
  encPub: Uint8Array;
}

export const SALT_SIZE = 16;
export const KEY_BUNDLE_SIZE = NONCE_SIZE + 64 + TAG_SIZE; // 104
export const DEFAULT_KDF: KdfParams = { t: 3, m: 64 * 1024, p: 1 };
const BUNDLE_AAD = utf8Encode('msgr-keybundle-v1');

export function keysFromSecrets(signSeed: Uint8Array, encPriv: Uint8Array): AccountKeys {
  if (signSeed.length !== 32 || encPriv.length !== 32) throw new Error('secrets must be 32 bytes');
  return { signSeed, signPub: ed25519Public(signSeed), encPriv, encPub: x25519Public(encPriv) };
}

export function generateKeys(): AccountKeys {
  return keysFromSecrets(randomBytes(32), randomBytes(32));
}

/** Argon2id over the password: authKey goes to the server, encKey stays local. */
export async function deriveKeys(
  password: string,
  salt: Uint8Array,
  params: KdfParams = DEFAULT_KDF,
): Promise<{ authKey: Uint8Array; encKey: Uint8Array }> {
  const out = await argon2idHash(password, salt, params, 64);
  return { authKey: out.slice(0, 32), encKey: out.slice(32, 64) };
}

export function sealKeyBundle(encKey: Uint8Array, nonce: Uint8Array, keys: AccountKeys): Uint8Array {
  return concat(nonce, aeadEncrypt(encKey, nonce, BUNDLE_AAD, concat(keys.signSeed, keys.encPriv)));
}

export function newKeyBundle(encKey: Uint8Array, keys: AccountKeys): Uint8Array {
  return sealKeyBundle(encKey, randomBytes(NONCE_SIZE), keys);
}

/** Throws when the bundle cannot be opened or does not match the public keys. */
export function openKeyBundle(encKey: Uint8Array, bundle: Uint8Array, signPub: Uint8Array, encPub: Uint8Array): AccountKeys {
  if (bundle.length !== KEY_BUNDLE_SIZE) throw new Error('malformed key bundle');
  const secrets = aeadDecrypt(encKey, bundle.subarray(0, NONCE_SIZE), BUNDLE_AAD, bundle.subarray(NONCE_SIZE));
  const keys = keysFromSecrets(secrets.slice(0, 32), secrets.slice(32, 64));
  if (!equal(keys.signPub, signPub) || !equal(keys.encPub, encPub)) throw new Error('key bundle does not match the public keys');
  return keys;
}

/**
 * Password change without the old password (PROTOCOL.md §3.2). A signed-in
 * device re-encrypts the key bundle from the account secrets it already has,
 * and proves it may do so by signing a server-issued challenge together with
 * the new material, instead of knowing the old password.
 */
export const PW_CHANGE_CHALLENGE_SIZE = 32;
const PW_CHANGE_PREFIX = utf8Encode('msgr-pwchange-v1');

/**
 * The bytes to sign:
 * "msgr-pwchange-v1" || challenge(32) || account_id(16) || device_id(16) ||
 * new_salt(16) || new_auth_key(32) || new_key_bundle(104) || sign_out_others(1).
 * Every field is fixed size, so the concatenation is unambiguous.
 */
export function passwordChangeMessage(
  challenge: Uint8Array,
  accountId: Uint8Array,
  deviceId: Uint8Array,
  newSalt: Uint8Array,
  newAuthKey: Uint8Array,
  newKeyBundle: Uint8Array,
  signOutOthers: boolean,
): Uint8Array {
  if (
    challenge.length !== PW_CHANGE_CHALLENGE_SIZE ||
    accountId.length !== 16 ||
    deviceId.length !== 16 ||
    newSalt.length !== SALT_SIZE ||
    newAuthKey.length !== 32 ||
    newKeyBundle.length !== KEY_BUNDLE_SIZE
  ) {
    throw new Error('invalid password change input');
  }
  return concat(PW_CHANGE_PREFIX, challenge, accountId, deviceId, newSalt, newAuthKey, newKeyBundle, new Uint8Array([signOutOthers ? 1 : 0]));
}

export function signPasswordChange(
  signSeed: Uint8Array,
  challenge: Uint8Array,
  accountId: Uint8Array,
  deviceId: Uint8Array,
  newSalt: Uint8Array,
  newAuthKey: Uint8Array,
  newKeyBundle: Uint8Array,
  signOutOthers: boolean,
): Uint8Array {
  return sign(signSeed, passwordChangeMessage(challenge, accountId, deviceId, newSalt, newAuthKey, newKeyBundle, signOutOthers));
}

File: app/lib/crypto/account.dart
// Account keys, password-derived keys and the encrypted key bundle,
// PROTOCOL.md §3.
import 'dart:typed_data';

import 'bytes.dart';
import 'primitives.dart';

const int saltSize = 16;
const int keyBundleSize = nonceSize + 64 + tagSize; // 104
const KdfParams defaultKdf = KdfParams(t: 3, m: 64 * 1024, p: 1);
final Uint8List _bundleAad = utf8Encode('msgr-keybundle-v1');

class AccountKeys {
  const AccountKeys({required this.signSeed, required this.signPub, required this.encPriv, required this.encPub});

  final Uint8List signSeed;
  final Uint8List signPub;
  final Uint8List encPriv;
  final Uint8List encPub;
}

class DerivedKeys {
  const DerivedKeys({required this.authKey, required this.encKey});

  /// Sent to the server as the login secret.
  final Uint8List authKey;

  /// Never leaves the device; unlocks the key bundle.
  final Uint8List encKey;
}

Future<AccountKeys> keysFromSecrets(List<int> signSeed, List<int> encPriv) async {
  if (signSeed.length != 32 || encPriv.length != 32) throw ArgumentError('secrets must be 32 bytes');
  return AccountKeys(
    signSeed: Uint8List.fromList(signSeed),
    signPub: await ed25519Public(signSeed),
    encPriv: Uint8List.fromList(encPriv),
    encPub: await x25519Public(encPriv),
  );
}

Future<AccountKeys> generateKeys() => keysFromSecrets(randomBytes(32), randomBytes(32));

Future<DerivedKeys> deriveKeys(String password, List<int> salt, [KdfParams params = defaultKdf]) async {
  final out = await argon2idHash(password, salt, params, 64);
  return DerivedKeys(authKey: out.sublist(0, 32), encKey: out.sublist(32, 64));
}

Future<Uint8List> sealKeyBundle(List<int> encKey, List<int> nonce, AccountKeys keys) async =>
    concat([nonce, await aeadEncrypt(encKey, nonce, _bundleAad, concat([keys.signSeed, keys.encPriv]))]);

Future<Uint8List> newKeyBundle(List<int> encKey, AccountKeys keys) => sealKeyBundle(encKey, randomBytes(nonceSize), keys);

/// Throws when the bundle cannot be opened or does not match the public keys.
Future<AccountKeys> openKeyBundle(List<int> encKey, List<int> bundle, List<int> signPub, List<int> encPub) async {
  if (bundle.length != keyBundleSize) throw const FormatException('malformed key bundle');
  final secrets = await aeadDecrypt(encKey, bundle.sublist(0, nonceSize), _bundleAad, bundle.sublist(nonceSize));
  final keys = await keysFromSecrets(secrets.sublist(0, 32), secrets.sublist(32, 64));
  if (!bytesEqual(keys.signPub, signPub) || !bytesEqual(keys.encPub, encPub)) {
    throw StateError('key bundle does not match the public keys');
  }
  return keys;
}

/// Password change without the old password (PROTOCOL.md §3.2). A signed-in
/// device re-encrypts the key bundle from the account secrets it already
/// holds, and proves it may do so by signing a server-issued challenge
/// together with the new material, instead of knowing the old password.
const int pwChangeChallengeSize = 32;
final Uint8List _pwChangePrefix = utf8Encode('msgr-pwchange-v1');

/// The bytes to sign:
/// "msgr-pwchange-v1" || challenge(32) || account_id(16) || device_id(16) ||
/// new_salt(16) || new_auth_key(32) || new_key_bundle(104) || sign_out_others(1).
/// Every field is fixed size, so the concatenation is unambiguous.
Uint8List passwordChangeMessage(
  List<int> challenge,
  List<int> accountId,
  List<int> deviceId,
  List<int> newSalt,
  List<int> newAuthKey,
  List<int> newKeyBundle, {
  required bool signOutOthers,
}) {
  if (challenge.length != pwChangeChallengeSize ||
      accountId.length != 16 ||
      deviceId.length != 16 ||
      newSalt.length != saltSize ||
      newAuthKey.length != 32 ||
      newKeyBundle.length != keyBundleSize) {
    throw ArgumentError('invalid password change input');
  }
  return concat([
    _pwChangePrefix,
    challenge,
    accountId,
    deviceId,
    newSalt,
    newAuthKey,
    newKeyBundle,
    [signOutOthers ? 1 : 0],
  ]);
}

Future<Uint8List> signPasswordChange(
  List<int> signSeed,
  List<int> challenge,
  List<int> accountId,
  List<int> deviceId,
  List<int> newSalt,
  List<int> newAuthKey,
  List<int> newKeyBundle, {
  required bool signOutOthers,
}) =>
    sign(
      signSeed,
      passwordChangeMessage(challenge, accountId, deviceId, newSalt, newAuthKey, newKeyBundle, signOutOthers: signOutOthers),
    );
